src/library_webgl.js

```diff
@@ -3834,7 +3834,7 @@ var LibraryGL = {
 
     if (!(mapping.access & 0x10)) /* GL_MAP_FLUSH_EXPLICIT_BIT */
       if (GL.currentContext.version >= 2) { // WebGL 2 provides new garbage-free entry points to call to WebGL. Use those always when possible.
-        GLctx.bufferSubData(target, mapping.offset, HEAPU8, mapping.mem, mapping.length);
+        if (mapping.length) GLctx.bufferSubData(target, mapping.offset, HEAPU8, mapping.mem, mapping.length);
```

Comment:
Likewise for these - could the spec enforce the proper behavior?

Reply:
See above. The specification clearly defines the behavior in JavaScript, which differs from the C++ versions of these APIs.